src/wire/bulk_report.rs

```diff
@@ -0,0 +1,349 @@
+    /// does not match.
+    pub fn check_fingerprint(&self) -> Result<(), BulkReportWireError> {
+        if self.fingerprint() != Self::FINGERPRINT {
+            Err(BulkReportWireError::InvalidFingerprint)
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Ensure the payload size is sufficient to hold bytes according to the header
+    /// fields `n_log_bytes` and `n_extension_bytes`.
+    ///
+    /// Returns `Err(BulkReportWireError::IncompletePayload)` if the buffer
+    /// is too short.
+    pub fn check_payload_len(&self) -> Result<(), BulkReportWireError> {
+        let n_log_bytes = self.n_log_bytes() as usize;
+        let n_extension_bytes = self.n_extension_bytes() as usize;
+        let len = self.buffer.as_ref().len();
+        if len < (field::PAYLOAD.start + n_log_bytes + n_extension_bytes) {
+            Err(BulkReportWireError::IncompletePayload)
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Consumes the bulk report, returning the underlying buffer
+    pub fn into_inner(self) -> T {
+        self.buffer
+    }
+
+    /// Return the length of a bulk report header
+    pub fn header_len() -> usize {
+        field::PAYLOAD.start
+    }
+
+    /// Return the length of a buffer required to hold a bulk report
+    /// with a payload length of `n_log_bytes` + `n_extension_bytes`
+    pub fn buffer_len(n_log_bytes: usize, n_extension_bytes: usize) -> usize {
+        field::PAYLOAD.start + n_log_bytes + n_extension_bytes
+    }
+
+    /// Return the length of the bulk report payload
+    pub fn payload_len(&self) -> usize {
+        let n_log_bytes = self.n_log_bytes() as usize;
+        let n_extension_bytes = self.n_extension_bytes() as usize;
+        n_log_bytes + n_extension_bytes
+    }
+
+    /// Return the `fingerprint` field
+    #[inline]
+    pub fn fingerprint(&self) -> u32 {
+        let data = self.buffer.as_ref();
+        LittleEndian::read_u32(&data[field::FINGERPRINT])
+    }
+
+    /// Return the `probe_id` field
+    #[inline]
+    pub fn probe_id(&self) -> u32 {
+        let data = self.buffer.as_ref();
+        LittleEndian::read_u32(&data[field::PROBE_ID])
+    }
+
+    /// Return the `n_log_bytes` field
+    #[inline]
+    pub fn n_log_bytes(&self) -> u32 {
+        let data = self.buffer.as_ref();
+        LittleEndian::read_u32(&data[field::N_LOG_BYTES])
+    }
+
+    /// Return the `n_extension_bytes` field
+    #[inline]
+    pub fn n_extension_bytes(&self) -> u32 {
+        let data = self.buffer.as_ref();
+        LittleEndian::read_u32(&data[field::N_EXT_BYTES])
+    }
+}
+
+impl<'a, T: AsRef<[u8]> + ?Sized> BulkReport<&'a T> {
+    /// Return a pointer to the payload
+    #[inline]
+    pub fn payload(&self) -> &'a [u8] {
+        let data = self.buffer.as_ref();
+        &data[field::PAYLOAD]
+    }
+}
+
+impl<T: AsRef<[u8]> + AsMut<[u8]>> BulkReport<T> {
+    /// Set the `fingerprint` field
+    #[inline]
+    pub fn set_fingerprint(&mut self, value: u32) {
```

Comment:
I don't think we have any reason to ever use a value other than FINGERPRINT, so I recommend we remove the `value` argument.